app/src/main/java/com/keylesspalace/tusky/EditProfileActivity.kt

```diff
@@ -94,93 +88,133 @@ class EditProfileActivity : BaseActivity(), Injectable {
-            override fun onResponse(call: Call<Account>, response: Response<Account>) {
-                if (!response.isSuccessful) {
-                    onAccountVerifyCredentialsFailed()
-                    return
-                }
-                val me = response.body()
-                oldDisplayName = me!!.displayName
-                oldNote = me.source?.note
-                oldLocked = me.locked
-
-                displayNameEditText.setText(oldDisplayName)
-                noteEditText.setText(oldNote)
-                lockedCheckBox.isChecked = oldLocked
-
-                if (!avatarChanged) {
-                    Picasso.with(avatarPreview.context)
-                            .load(me.avatar)
-                            .placeholder(R.drawable.avatar_default)
-                            .into(avatarPreview)
+                        displayNameEditText.setText(me.displayName)
+                        noteEditText.setText(me.source?.note)
+                        lockedCheckBox.isChecked = me.locked
+
+                        accountFieldEditAdapter.setFields(me.source?.fields ?: emptyList())
+
+                        if(viewModel.avatarData.value == null) {
+                            Picasso.with(this@EditProfileActivity)
+                                    .load(me.avatar)
+                                    .placeholder(R.drawable.avatar_default)
+                                    .into(avatarPreview)
+                        }
+
+                        if(viewModel.headerData.value == null) {
+                            Picasso.with(this@EditProfileActivity)
+                                    .load(me.header)
+                                    .into(headerPreview)
+                        }
+
+                    }
                 }
-                if (!headerChanged) {
-                    Picasso.with(headerPreview.context)
-                            .load(me.header)
-                            .into(headerPreview)
+                is Error -> {
+                    val snackbar = Snackbar.make(avatarButton, R.string.error_generic, Snackbar.LENGTH_LONG);
+                    snackbar.setAction(R.string.action_retry) {
+                        viewModel.obtainProfile()
+                    }
+                    snackbar.show()
+
                 }
             }
+        })
+
+        observeImage(viewModel.avatarData, avatarPreview, avatarProgressBar)
+        observeImage(viewModel.headerData, headerPreview, headerProgressBar)
 
-            override fun onFailure(call: Call<Account>, t: Throwable) {
-                onAccountVerifyCredentialsFailed()
+        viewModel.saveData.observe(this, Observer<Resource<Boolean>> {
+            when(it) {
+                is Success -> {
+                    if(it.data == true) {
+                        privatePreferences.edit()
+                                .putBoolean("refreshProfileHeader", true)
```

Comment:
1. Non-constant constant
2. Kinda weird solution IMO - couldn't we just send an event? Why do we have to save this?

Reply:
yeah I updated it so now we dont even need to reload the profile data because we get it back from the update call and pass it araound with events!